File: Binary Search/3049. Earliest Second to Mark Indices II.cpp
/*

需要first index: 因为想要尽早的set nums = 0 


a. Why backwards? 
Because after zeroing out a number you have to make sure that there is enough time to mark it as well. 
By going backwards we can keep track of the available time in the future.

当time_available==0 时候, 必须要drop 一个, time_available == 0 表示后面没有时间来heap中的点

最后heap 保留的是，最大number可以被zero out的
*/

class Solution {
public:
    bool isPossible(vector<int>& nums, vector<int>& changeIndices, unordered_map<int,int>& first_index_to_value_mp, int bound){
        int time_available = 0;
        vector<long>heap;
        for(int i = min(bound, (int)changeIndices.size()) - 1; i >= 0; --i){
            if (first_index_to_value_mp.count(i)) {
                heap.push_back(-nums[first_index_to_value_mp[i]-1]);
                //cout<<" [i push] "<<i<<" val "<<nums[first_index_to_value_mp[i]-1]<<" time_available "<<time_available<<endl;
                push_heap(heap.begin(), heap.end());

                if (time_available > 0) {
                    --time_available ;
                } else {
                    time_available += 1;
                    //cout<<" i pop "<<i<<" val "<<nums[first_index_to_value_mp[i]-1]<<endl;
                    pop_heap(heap.begin(), heap.end());
                    heap.pop_back();
                    //pop_heap does not remove elements from the container. It can't, since it doesn't even have access to the container
                }
            } else {
                ++time_available;
            }
        }
        long nums_total = accumulate(nums.begin(), nums.end(), 0L);
        long q_total = -accumulate(heap.begin(), heap.end(), 0L);
        long decrement_total = nums_total - q_total + nums.size() - heap.size();
        long mark_zero = heap.size() * 2;
        //cout<<" decrement_total "<<decrement_total <<" nums_total "<< nums_total <<" q_total "<<q_total<<" heap.size() " << heap.size()<<" mark_zero "<<mark_zero<<endl;
        return decrement_total + mark_zero <= bound;
    }


    int earliestSecondToMarkIndices(vector<int>& nums, vector<int>& changeIndices) {
        unordered_map<int, int>first_index, first_index_to_value_mp;
        int m = changeIndices.size();
        for(int i = 0; i <m; ++i){
            if (first_index.count(changeIndices[i]) == 0 && nums[changeIndices[i]-1]!=0){
                first_index[changeIndices[i]] = i;
            }
        }
        for(auto it: first_index){
            first_index_to_value_mp[it.second] = it.first;
        }
        int low =0 , high = m + 1;
        while (low < high){
            int mid = (low + high)/2;
            if (isPossible(nums, changeIndices, first_index_to_value_mp, mid)){
                high = mid;
                //cout<<" low "<<low <<" high "<<high << " mid "<<mid<<" possible "<<endl;
            } else {
                //cout<<" low "<<low <<" high "<<high << " mid "<<mid<<" not possible "<<endl;
                low = mid + 1;
            }
        }
        return low > m? -1: low;

    }
};

File: Binary Search/3049. Earliest Second to Mark Indices II.py
"""

需要first index: 因为想要尽早的set nums = 0 


a. Why backwards? 
Because after zeroing out a number you have to make sure that there is enough time to mark it as well. 
By going backwards we can keep track of the available time in the future.

当time_available==0 时候, 必须要drop 一个, time_available == 0 表示后面没有时间来heap中的点

最后heap 保留的是，最大number可以被zero out的

"""
class Solution:
    def earliestSecondToMarkIndices(self, nums: List[int], changeIndices: List[int]) -> int:
        first_index = defaultdict(int)
        for i, v in enumerate(changeIndices):
            if v not in first_index and nums[v-1]!=0:
                #如果v 是0, mark的机会被这里算  len(nums) - len(q)
                first_index[v] = i 

        first_index_to_value = {i:v for v, i in first_index.items()}
        n, m = len(nums), len(changeIndices)
        print("first_index_to_value ",first_index_to_value)
        def possible(bound):
            q = []
            time_available = 0
            for i in range(min(bound,m)-1, -1, -1):
                if i in first_index_to_value:
                    heapq.heappush(q, nums[first_index_to_value[i]-1])
                    if time_available:
                        time_available -= 1 
                    else:
                        time_available += 1 
                        heapq.heappop(q) #因为把之前的push value清理走了, 加一个时间
            
                else:
                    time_available += 1 
            
            # sum(nums) - sums(q) different between target and can be set as 0 
            # len(nums) - lens(q): time required to mark value 
            decrement_total = sum(nums) - sum(q) + len(nums) - len(q)
            mark_as_zero = len(q) * 2 # set 一次，mark 一次
            #print("q ", q, "decrement_total", decrement_total, "mark_as_zero", mark_as_zero)
            return decrement_total + mark_as_zero <= bound

        low, high = 0, m+1
        while low < high:
            mid = (low + high)//2
            if possible(mid):
                high = mid
            else:
                low = mid + 1 
        return low if low <= m else -1
         decrement_total 3 nums_total 4 q_total 2 heap.size() 1 mark_zero 2
